### PR TITLE
magento/devdocs#: Docs: Create topic for addDownloadableProductsToCart mutation

### PR DESCRIPTION
**Pull request by atwixfirster**
## Purpose of this pull request

This pull request (PR) fixes https://github.com/magento/graphql-ce/issues/1003 issue.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.3/graphql/mutations/add-downloadable-products.html

## Links to Magento source code

- https://github.com/magento/graphql-ce/pull/412

CC: @keharper 

Thank you!

whatsnew
Added the [addDownloadableProductsToCart](https://devdocs.magento.com/guides/v2.3/graphql/mutations/add-downloadable-products.html) mutation.

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
+---
+group: graphql
+title: addDownloadableProductsToCart mutation
+contributor_name: Atwix
+contributor_link: https://www.atwix.com/
+---
+
+A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide  customizable options.
```
Comment (rogyar) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide customizable options.
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+The `AddDownloadableProductsToCartInput` object must contain the following attributes:
+
+Attribute |  Data Type | Description
+--- | --- | ---
+`cart_id` | String! | The unique ID that identifies the customer's cart
+`cart_items` | [DownloadableProductCartItemInput!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item
+
+### DownloadableProductCartItemInput object {#DownloadableProductCartItemInput}
+
+The `DownloadableProductCartItemInput` object must contain the following attributes:
```
Comment (rogyar) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]: We say that the object must contain the following attributes, however, `downloadable_product_links` and `customizable_options` are optional attributes. So "must" is not the 100% correct assumption here.

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+Attribute |  Data Type | Description
+--- | --- | ---
+`customizable_options` |[[CustomizableOptionInput]](#CustomizableOptionInputVirtual) | An array that defines customizable options for the product
```
Comment (rogyar) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
`customizable_options` |[[CustomizableOptionInput!]](#CustomizableOptionInputVirtual) | An array that defines customizable options for the product
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+The top-level `AddDownloadableProductsToCartInput` object is listed first. All child objects are listed in alphabetical order.
+
+### AddDownloadableProductsToCartInput object {#AddDownloadableProductsToCartInput}
+
+The `AddDownloadableProductsToCartInput` object must contain the following attributes:
+
+Attribute |  Data Type | Description
+--- | --- | ---
+`cart_id` | String! | The unique ID that identifies the customer's cart
+`cart_items` | [DownloadableProductCartItemInput!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item
```
Comment (rogyar) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
`cart_items` | [[DownloadableProductCartItemInput!]!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 3 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **fixed**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 2 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: Really a good point.

**fixed**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 4 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **fixed**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 1 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: Nice catch :)

**fixed**

**Comment by atwixfirster** [comment]
@rogyar , all your suggestions have been implemented. Could you please review?

Many thanks as always

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+These examples show how to add a downloadable product into shopping cart with enabled/disabled `Links can be purchased separately` downloadable product option.
+
+### Add a downloadable product to a cart with enabled `Links can be purchased separately`
+
+The following example displays how to add a downloadable product with `sku` value of `240-LV09` and with the custom downloadable links: `Episode 2`, `Episode 3` from the Luma sample data into shopping cart. The product with `sku` `240-LV09` has enabled `Links can be purchased separately` option, so you must specify downloadable links which will be added into cart.
+
+**Request**
+
+```text
```
Comment (lenaorobei) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
```graphql
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+title: addDownloadableProductsToCart mutation
+contributor_name: Atwix
+contributor_link: https://www.atwix.com/
+---
+
+A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide customizable options.
+
+## Syntax
+
+`mutation: {addDownloadableProductsToCart(input: AddDownloadableProductsToCartInput): {AddDownloadableProductsToCartOutput}}`
```
Comment (lenaorobei) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]: Suggest using multi-line `graphql` representation.

Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 5 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

**Comment by atwixfirster** [comment]
@lenaorobei , could you please review?

Thanks

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        ]
+      }
+    }
+  }
+}
+```
+
+### Add a downloadable product to a cart with disabled `Links can be purchased separately`
+
+The following example displays how to add a downloadable product with the `sku` value of `240-LV07` from the Luma sample data into shopping cart. This product has disabled `Links can be purchased separately` option, so all downloadble links will be added into cart automatically.
```
Comment (keharper) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
The following example shows how to add a downloadable product in which the **Links can be purchased separately** option is disabled. All downloadable links are added to the cart automatically.
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+### AddDownloadableProductsToCartInput object {#AddDownloadableProductsToCartInput}
+
+The `AddDownloadableProductsToCartInput` object must contain the following attributes:
+
+Attribute |  Data Type | Description
+--- | --- | ---
+`cart_id` | String! | The unique ID that identifies the customer's cart
+`cart_items` | [[DownloadableProductCartItemInput!]!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item
+
+### DownloadableProductCartItemInput object {#DownloadableProductCartItemInput}
```
Comment (keharper) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]: This section is not listed alphabetically.

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+### Cart object {#CartObject}
+
+{% include graphql/cart-object.md %}
+
+## Errors
+
+Error | Description
+--- | ---
+`Required parameter "cart_id" is missing` | Mutation `cart_id` parameter is empty
```
Comment (keharper) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
`Required parameter "cart_id" is missing` | The mutation does not contain a `cart_id` parameter.
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+`Required parameter "cart_items" is missing` | Mutation `cart_items` parameter is empty or is not a type of `array`
```
Comment (keharper) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
`Required parameter "cart_items" is missing` | The `cart_items` parameter is empty or is not of type `array`.
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+`Please specify product link(s).` | Error appears when you try to add a downloadable product with enabled `Links can be purchased separately` option and you did not specify links which you want to buy in the `mutation` query.
```
Comment (keharper) [inline guides/v2.3/graphql/mutations/add-downloadable-products.md]:
```suggestion
`Please specify product link(s).` | You tried to add a downloadable product in which the `Links can be purchased separately` option is enabled, but you did not specify individual product links.
```

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 8 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **fixed**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 12]
```diff
@@ -0,0 +1,266 @@
+---
+group: graphql
+title: addDownloadableProductsToCart mutation
+contributor_name: Atwix
+contributor_link: https://www.atwix.com/
+---
+
+A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide customizable options.
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        input: AddDownloadableProductsToCartInput
+    ): {
+        AddDownloadableProductsToCartOutput
+    }
+}
+```
+
+## Example usage
+
+These examples show how to add a downloadable product into shopping cart with enabled/disabled `Links can be purchased separately` downloadable product option.
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+### Add a downloadable product to a cart with enabled `Links can be purchased separately`
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+
+The following example displays how to add a downloadable product with `sku` value of `240-LV09` and with the custom downloadable links: `Episode 2`, `Episode 3` from the Luma sample data into shopping cart. The product with `sku` `240-LV09` has enabled `Links can be purchased separately` option, so you must specify downloadable links which will be added into cart.
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 7 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 9 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 10 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md (hunk 11 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **implemented**

**Comment by atwixfirster** [comment]
@keharper , all your suggestions have been implemented. Could you please review?

Thanks!

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+### CustomizableOptionInput object {#CustomizableOptionInputVirtual}
+
+The `CustomizableOptionInput` object must contain the following attributes:
+
+{% include graphql/customizable-option-input.md %}
+
+### DownloadableProductCartItemInput object {#DownloadableProductCartItemInput}
+
+The `DownloadableProductCartItemInput` object must contain the following attribute:
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **fixed**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+Attribute |  Data Type | Description
+--- | --- | ---
+`data` | [CartItemInput!](#CartItemInputVirtual) | An object containing the `sku` and `quantity` of the product
+
+and the next optional ones:
+
+Attribute |  Data Type | Description
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **done**

File: guides/v2.3/graphql/mutations/add-downloadable-products.md [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+--- | --- | ---
+`customizable_options` |[[CustomizableOptionInput!]](#CustomizableOptionInputVirtual) | An array that defines customizable options for the product
+`downloadable_product_links` | [[DownloadableProductLinksInput!]](#DownloadableProductLinksInput) | An object containing the `link_id` of the downloadable product link
+
+### DownloadableProductLinksInput object {#DownloadableProductLinksInput}
+
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/mutations/add-downloadable-products.md]: **added**

**Comment by atwixfirster** [comment]
@keharper ,

**done** :)

Could you please review again? Thanks

**Comment by keharper** [comment]
running tests

**Comment by keharper** [comment]
@atwixfirster The linting checks are failing. Please remove the extra spaces at the end of line 48. Also, remove the comma after line 39.

**Comment by atwixfirster** [comment]
> @atwixfirster The linting checks are failing. Please remove the extra spaces at the end of line 48. Also, remove the comma after line 39.

**done**

**Comment by keharper** [comment]
running tests

**Comment by keharper** [comment]
Another new topic from @atwixfirster. Thanks!

**Comment by atwixfirster** [comment]
> Another new topic from @atwixfirster. Thanks!

With my pleasure, @keharper ! 

Thank you!

**Pull request merged**